contracts/contracts/harvest/BaseHarvester.sol

```diff
@@ -12,37 +12,63 @@ import { IVault } from "../interfaces/IVault.sol";
 import { IOracle } from "../interfaces/IOracle.sol";
 import { IStrategy } from "../interfaces/IStrategy.sol";
 import { IUniswapV2Router } from "../interfaces/uniswap/IUniswapV2Router02.sol";
+import { IUniswapV3Router } from "../interfaces/uniswap/IUniswapV3Router.sol";
+import { IBalancerVault } from "../interfaces/balancer/IBalancerVault.sol";
+import { ICurvePool } from "../strategies/ICurvePool.sol";
 import "../utils/Helpers.sol";
 
 abstract contract BaseHarvester is Governable {
     using SafeERC20 for IERC20;
     using SafeMath for uint256;
     using StableMath for uint256;
 
-    event UniswapUpdated(address _address);
-    event SupportedStrategyUpdate(address _address, bool _isSupported);
+    enum SwapPlatform {
+        UniswapV2Compatible,
+        UniswapV3,
+        Balancer,
+        Curve
+    }
+
+    event SupportedStrategyUpdate(address strategyAddress, bool isSupported);
     event RewardTokenConfigUpdated(
-        address _tokenAddress,
-        uint16 _allowedSlippageBps,
-        uint16 _harvestRewardBps,
-        address _uniswapV2CompatibleAddr,
-        uint256 _liquidationLimit,
-        bool _doSwapRewardToken
+        address tokenAddress,
+        uint16 allowedSlippageBps,
+        uint16 harvestRewardBps,
+        SwapPlatform platform,
+        address swapRouterAddr,
+        bytes swapData,
+        uint256 liquidationLimit,
+        bool doSwapRewardToken
+    );
+    event RewardTokenSwapped(
+        address indexed rewardToken,
+        address indexed swappedInto,
+        SwapPlatform platform,
+        uint256 amountIn,
+        uint256 amountOut
+    );
+    event RewardProceedsTransferred(
+        address indexed token,
+        address farmer,
+        uint256 protocolShare,
+        uint256 farmerShare
     );
+    event RewardProceedsAddressChanged(address newProceedsAddress);
 
     // Configuration properties for harvesting logic of reward tokens
     struct RewardTokenConfig {
         // Max allowed slippage when swapping reward token for a stablecoin denominated in basis points.
         uint16 allowedSlippageBps;
         // Reward when calling a harvest function denominated in basis points.
         uint16 harvestRewardBps;
-        /* Address of Uniswap V2 compatible exchange (Uniswap V2, SushiSwap).
-         */
-        address uniswapV2CompatibleAddr;
+        // Address of compatible exchange router (Uniswap V2, SushiSwap).
+        address swapRouterAddr;
         /* When true the reward token is being swapped. In a need of (temporarily) disabling the swapping of
          * a reward token this needs to be set to false.
          */
         bool doSwapRewardToken;
+        // Platform to use for Swapping
+        SwapPlatform platform;
```

Comment:
good call adding SwapPlatform next to a bool. Since a bool and uint8 (underlying enum implementation) will be packed together

Reply:
nit: we could do gas optimisation and pack `doSwapRewardToken` & `platform` next to the 2 `uint16` at the top of the struct. To save on storage reads. But that would require updating reward token config of all of the tokens. Which already might be the case will all these additions?

Reply:
Isn't it already in a single slot? `uint16,uint16,address,bool` would use `16 + 16 + 160 + 8` bits, so 200 bits (or 25 bytes), which is still under one slot.

I thought of changing the order of the elements at first as well, but we check the old router address and remove any allowance when using `setRewardTokenConfig` method. Changing this struct would mean that we might potentially cause it to fail on mainnet after the upgrade since the value of the old router address might not be `address(0)` or the right one due to change in slot